### ISSUE TITLE
early injection fails on non-PIE-enabled kernels or when ASLR is disabled

### ISSUE THREAD

**Issue by algr** [DynamoRIO/dynamorio#1708]
My build of DR for 32-bit ARM is failing (no client, any program) with this message:
```
ERROR: run 'set disable-randomization off' to run from gdb
```

I don't understand the message or the comment in loader.c.  The code seems to be trying to detect if ASLR is enabled and fails if so.  But if ASLR is the problem, then gdb's default disabling of ASLR is fine.  The problem would be if the user had done 'set disable-randomization off' to re-enable ASLR.  And then the solution is not to do that, or to do 'set disable-randomization on'.

But the problem here is that we get the fail even when not running under gdb and when ASLR is disabled system-wide (set /proc/sys/kernel/randomize_va_space to 0).

Judging by the test, whichever library this is in has been loaded at 0x5xxxxxxx and for some reason that's a problem.

**Comment by derekbruening** [comment]
This is problem worried about in the past.  If the user sets /proc/sys/kernel/randomize_va_space to 0, or is on a kernel that doesn't support PIE, the exec of libdynamorio.so will place it at a fixed location computed from the highest user address.

We have two solutions for supporting this:

1) Create a separate libdynamorio.so with a different preferred address for each possible fixed location.  There should only be 3 of these in common use for 32-bit (2GB, 3GB, and 4GB max) and a similar handful for 64-bit.  The fixed location should not vary across a fork, so we can have drrun pick the build and have from-parent injection use the current build.

2) Switch back to exec of an ET_EXEC binary.

**Comment by derekbruening** [comment]
For any users hitting this (must be using a custom build as the last official release did not have early injection by default), run with "-late" to drrun, or you can set the preferred address to the particular fixed location your setup is using (`cmake -Dpreferred_base=0x55555554`).

**Comment by derekbruening** [comment]
3) Have drrun auto-detect if PIE + ASLR are enabled (not sure how for PIE)
   and fall back to late injection if not.

**Comment by derekbruening** [comment]
Xref #47, #1676

**Comment by algr** [comment]
This is using a 4.1 kernel for 32-bit ARM, with ASLR enabled.  Running "cat /proc/self/maps" a few times shows that the system dynamic libraries are being relocated to different places each time.  So it would seem that PIE + ASLR is enabled.

I was confused by the comment
```
    /* i#1676: try to detect ALSR which happens if we're launched from within
     * gdb w/o 'set disable-randomization off'.  We assume here that our
     * preferred base does not start with 0x5... and that the ASLR base does:
     *   64-bit: 0x555555b323ab
     *   32-bit: 0x56555000
     */
```

Isn't it more correct to say that ASLR happens (when enabled at system level) if we're
(a) not launched from within gdb 
(b) launched from gdb _with_ 'set disable-randomization off'?

Looking at libdynamorio.so, the preferred base is around 0x71000000.

**Comment by algr** [comment]
Adding some tracing shows that:

 - with ASLR enabled, the module is loaded at a range of addresses starting 0x5xxxxxxx

 - with ASLR disabled, the module is loaded at a single address starting 0x5xxxxxxx

So, ASLR has no effect on whether the ASLR test fails.  It always fails, since the address is always 0x5xxxxxxx.  Presumably, changing the preferred base address to 0x5xxxxxxx would only make that more likely.  Maybe the answer is to change the preferred base address to 0x6xxxxxxx?

Adding -late seems to work around the problem.

**Comment by derekbruening** [comment]
> Running "cat /proc/self/maps" a few times shows that the system dynamic libraries are being relocated
> to different places each time. So it would seem that PIE + ASLR is enabled.

Technically that tells you nothing about whether PIE is enabled.

Yes, the comment wording is slightly messed up (s/detect ASLR/detect lack of ASLR/ and s/ASLR base/non-ASLR base/).

> Maybe the answer is to change the preferred base address to 0x6xxxxxxx?

No, that will not solve anything.  We have already enumerated the solutions above.

**Comment by derekbruening** [comment]
Is it typical to run with ASLR disabled on your system?  Is this just for easier debugging and reproducability?

> It always fails, since the address is always 0x5xxxxxxx.

Are you saying that your kernel execs libdynamorio at an 0x5* address when ASLR is enabled?  Are you sure that's a default-built libdynamorio with a non-zero preferred base?  Is this a custom kernel?  Are you sure this is on exec and not just on a regular load?

**Comment by algr** [comment]
The ASLR test fails every time, so &privload_early_inject == 0x5xxxxxxx... I put in a hex readout of its actual address, just to make sure.  The build is default, I pulled from github today and rebuilt.

It looks like the library preferred base is 0x71000000:
```
root@sdperf-ve2:~/dynamorio# readelf -l ../dynamorio/build/lib32/release/libdynamorio.so

Elf file type is DYN (Shared object file)
Entry point 0x710b327c
There are 6 program headers, starting at offset 52

Program Headers:
  Type           Offset   VirtAddr   PhysAddr   FileSiz MemSiz  Flg Align
  LOAD           0x000000 0x71000000 0x71000000 0xda7a4 0xda7a4 R E 0x8000
  LOAD           0x0da978 0x710e2978 0x710e2978 0x526b8 0x6be08 RW  0x8000
  DYNAMIC        0x117f60 0x7111ff60 0x7111ff60 0x000a0 0x000a0 RW  0x4
  NOTE           0x0000f4 0x710000f4 0x710000f4 0x00024 0x00024 R   0x4
  GNU_STACK      0x000000 0x00000000 0x00000000 0x00000 0x00000 RW  0x10
  GNU_RELRO      0x0da978 0x710e2978 0x710e2978 0x3d688 0x3d688 R   0x1

root@sdperf-ve2:~/dynamorio# readelf -d ../dynamorio/build/lib32/release/libdynamorio.so | more

Dynamic section at offset 0x117f60 contains 16 entries:
  Tag        Type                         Name/Value
 0x0000000e (SONAME)                     Library soname: [libdynamorio.so.5.0]
 0x00000010 (SYMBOLIC)                   0x0
 0x0000000c (INIT)                       0x710b43bd
 0x00000004 (HASH)                       0x71000118
 0x6ffffef5 (GNU_HASH)                   0x710013c0
 0x00000005 (STRTAB)                     0x71005254
 0x00000006 (SYMTAB)                     0x71002864
 0x0000000a (STRSZ)                      13546 (bytes)
 0x0000000b (SYMENT)                     16 (bytes)
 0x00000011 (REL)                        0x71008740
 0x00000012 (RELSZ)                      105640 (bytes)
 0x00000013 (RELENT)                     8 (bytes)
 0x00000018 (BIND_NOW)                   
 0x6ffffffb (FLAGS_1)                    Flags: NOW
 0x6ffffffa (RELCOUNT)                   13205
 0x00000000 (NULL)                       0x0
```

I don't really understand the question about it being on exec or a regular load... all I'm doing is starting DR with no options (the hex in the message below, is what I added):
```
root@sdperf-ve2:~# ./dr/build/bin32/drrun -- ls
ERROR: run 'set disable-randomization off' to run from gdb:54B6AE79
```

**Comment by derekbruening** [comment]
> I don't really understand the question about it being on exec or a regular load

Everything depends on the kernel's behavior on exec of ET_DYN (see comment above "handled strangely").

It sounds like your kernel either has no PIE support or has different PIE support from what we've seen.

Looking at recent kernel changes: indeed, April of this year there is a change that breaks our injection, if you have this in your kernel:

https://kernel.googlesource.com/pub/scm/linux/kernel/git/rafael/linux-pm/+/d1fd836dcf00d2028c700c7e44d2c23404062c90%5E!/

That rules out soln#1.  Looks like we'll have to go back to ET_EXEC.

**Comment by derekbruening** [comment]
The simplest solution is to have 2 builds of libdynamorio.so, one ET_EXEC for early injection and the other ET_DYN for legacy late injection.  Then we have very little to change in the rest of the code, except the cross-arch file picker.

**Comment by derekbruening** [comment]
For Android, we have two choices:

1. Build a bootstrap relocator that will work without itself being relocated.  We could then stick with ET_DYN with no PT_INTERP.

2. Have our code loaded and relocated using the Android loader and then take over and unmap the loader and undo any global state like segment allocation if necessary.  We would then have ET_DYN with PT_INTERP.

Either solution will also work on Linux, and on Linux we have a third solution:

3. Use an ET_EXEC initial binary, preferably libdynamorio.so (with either two builds or we abandon support for late injection).

The concern for solution 2 on Linux is an older kernel with no PIE support: we would have to confirm that PT_INTERP is checked for ET_DYN on all kernels we care about.

**Comment by zhaoqin** [comment]
Base on the code at http://lxr.free-electrons.com/source/fs/binfmt_elf.c, even the earliest version 2.0.40 (http://lxr.free-electrons.com/source/fs/binfmt_elf.c?v=2.0.40), PT_INTERP is checked for both ET_DYN and ET_EXEC.

**Comment by zhaoqin** [comment]
From the comment in the loader binutils/bfd/elflink.c, there is no .interp section in shared library at all, so we cannot create a shared library with ET_DYN and PT_INTERP exist at the same time.

**Comment by zhaoqin** [comment]
We can add PT_INTERP to the libdynamorio, however, it breaks some assumption on 32-bit x86 Linux:
in core/unix/os.c:choose_gdt_slots
```
        int index = SELECTOR_INDEX(os_tls->app_lib_tls_reg);
        if (index == 0) {
            /* An index of zero means the app has no TLS (yet), and happens
             * during early injection.  We use -1 to grab a new entry.  When the
             * app asks for its first table entry with set_thread_area, we give
             * it this one and emulate its usage of the segment.
             */
```
Because of having interp in libdynamorio, the interp will set up the TLS when loading libdynamorio, so even for early inject, we still have the app's TLS segment setup.

**Comment by zhaoqin** [comment]
We have an initial effort to add system linker to libdynamorio: https://codereview.appspot.com/258530043/.
It works in most cases, but have a few potential problems:
1. tls issue as discussed in the previous comment, we need clear the tls setup by libdynamorio's linker.
2. libdynamorio's interp breaks the assumption of `privload_early_inject` param `old_libdr_base` and `old_libdr_size` both being 0 in early injection.
3. relocating libdynamorio's dynamic section is messy. There are several situations that libdynamorio may or may not be relocated, e.g., no interp, not relocated, with interp, relocated, with interp but reloaded, not relocated. It is hard to tell which case we are in.
So we decide to implement our own bootstrap relocator instead.

**Comment by zhaoqin** [comment]
Closed by 0bc2b0b
